feat(tree): Generate JSON Schema from tree view schema (#21984)
microsoft/FluidFramework: .changeset/purple-analysts-depart.md

---
"@fluidframework/tree": minor
---

> Support generation of [JSON Schema](https://json-schema.org/docs) from Shared Tree view schema (alpha).
> 
> Adds canonical `JSON Schema` representation of Shared Tree schema, and adds `getJsonSchema` function for getting that representation for a given `TreeNodeSchema.

Reviewer: ```suggestion
Support generation of JSON Schema from Shared Tree view schema (alpha)

Adds alpha-quality support for canonical [JSON Schema](https://json-schema.org/docs) representation of Shared Tree schema and adds a `getJsonSchema` function for getting that representation for a given `TreeNodeSchema`.
```

> This `JSON Schema` representation can be used to describe schema requirements to external systems, and can be used with validation tools like [ajv](https://ajv.js.org/) to validate data before inserting it into a Shared Tree.

Reviewer: ```suggestion
This JSON Schema representation can be used to describe schema requirements to external systems, and can be used with validation tools like [ajv](https://ajv.js.org/) to validate data before inserting it into a Shared Tree.
```

Unless "JSON Schema" (with the space) is a code term, omit the code formatting.

> This `JSON Schema` representation can be used to describe schema requirements to external systems, and can be used with validation tools like [ajv](https://ajv.js.org/) to validate data before inserting it into a Shared Tree.

Reviewer: ```suggestion
This `JSON Schema` representation can be used to describe schema requirements to external systems and can be used with validation tools like [ajv](https://ajv.js.org/) to validate data before inserting it into a `SharedTree`.
```

> Example:

Reviewer: ```suggestion
### Example
```

Author: What is the "root" heading level of a changeset? I considered doing this, but it wasn't clear to me what the right level was (if any). Is this documented somewhere?

Author: (Updated to use your recommendation, but I would still love to see the heading level expectations for this documented somewhere if they aren't already).

Reviewer: There's only every supposed to be one H1, so the headings in the release notes start at H2. But we can't expect everyone to always get this right, so we will continue to have reviews and whatnot for release notes. And with the previews, it'll be obvious if someone gets it wrong.

Author: Then why was the recommendation to be a third-level heading? Should it be second-level? Or was the suggested level mostly arbitrary? If we're going to recommend devs use headings in their changelogs, I'd like for them to be consistent.

> A Shared Tree schema like...

Reviewer: ```suggestion
A Shared Tree schema like the following:
```

```typescript
class MyObject extends schemaFactory.object("MyObject", {
	foo: schemaFactory.number,
	bar: schemaFactory.optional(schemaFactory.string),
});
```

> ...will yield JSON Schema like...

Reviewer: ```suggestion
Such a SharedTree schema will yield JSON Schema like the following:
```

Author: I went with a different wording to keep the sentence flow going across the example. Let me know what you think.

```json
{
	"$defs": {
		"com.fluidframework.leaf.string": {
			"type": "string",
		},
		"com.fluidframework.leaf.number": {
			"type": "number",
		},
		"com.myapp.MyObject": {
			"type": "object",
			"properties": {
				"foo": { "$ref": "com.fluidframework.leaf.number" },
				"bar": { "$ref": "com.fluidframework.leaf.string" },
			},
			"required": ["foo"],
		},
	},
	"anyOf": [ { "$ref": "#/$defs/com.myapp.MyObject" } ],

> }

Reviewer 2: I'm assuming metadata will be a part of a future PR?

Author: The `com.fluidframework.leaf.foo` names are just identifiers, so I don't know how much they matter. The `type: foo` information is more meaningful, so I suspect the LLMs would pay more attention to that information over identifier names.

_But_ since everything is just pattern-matching for LLMs at the end of the day, I suspect you're right that common names like "string" could yield better output than `com.fluidframework.leaf.string`.

But I also didn't encounter any issues with the latter in my experimentation. Definitely an area for more experimentation in the future.

Author: And yes, schema metadata will be a follow-up. I think we have a concrete story for it now, so I'll probably brush off my design doc for it in the coming sprint.

Reviewer 2: More complex use cases, or highly nested objects will potentially have issues. It's hard for me to say things concretely here since we have no tests so I am only speaking from my own experience developing with LLMs, which is that you want to make things as close to the training data and JSON attribute names in the dot format are unusual.

That being said I don't think it's a blocker to this PR's work. It would probably be one more conversion layer if we added it.

Reviewer 3: Even if GPT doesn't care, it would also just cut down the size of the exported text. I wonder if we should actually replace "com.fluidframework.leaf.number" with "number" not just here but all across our system. It seems appropriate for the built-in primitives. The reason we don't do that already is because we don't want to collide with a user-defined type called "number". Most users will provide a scope, so it will be impossible to construct a scopeless type (i.e. you'd end up with "myscope.number", not "number") anyway. But it seems to me we should just ban the use of the built-in types. Or, we could reserve a special character that the user isn't allowed to use as part of a type name, and all our built-in types start with that character (that lets us safely add more primitives in the future). Probably a good discussion to have with Craig, who has thought about this already I'm sure.

> ```

Reviewer: ``````suggestion
```

> [!WARNING]
> This API is alpha quality and may change at any time.
``````

Since these are shown in GitHub ultimately it seems OK to use their aside formatting.

Might make sense earlier in the note rather than at the end. If we have some "here's what alpha means" docs we can link to them too.

---
"@fluidframework/tree": minor
---

> 

Reviewer 4: nit: is this missing a section? Or are sections not needed until 2.2?

Author: Ah, good catch. Updated.

Support generation of JSON Schema from Shared Tree view schema (alpha)

> [!WARNING]
> This API is [alpha quality](https://fluidframework.com/docs/build/releases-and-apitags/#api-support-levels) and may change at any time.

Adds alpha-quality support for canonical [JSON Schema](https://json-schema.org/docs) representation of Shared Tree schema and adds a `getJsonSchema` function for getting that representation for a given `TreeNodeSchema`.
This JSON Schema representation can be used to describe schema requirements to external systems, and can be used with validation tools like [ajv](https://ajv.js.org/) to validate data before inserting it into a `SharedTree`.

> ### Example

Author: ```suggestion
## Example
```
